### PR TITLE
OPF Model docstrings

### LINKED ISSUE #3588: Update OPF model docstrings

### PR DESCRIPTION
**Pull request by rhyolight**
fixes #3588

### CODE REVIEW

File: src/nupic/frameworks/opf/htm_prediction_model.py [hunk 1]
```diff
@@ -69,6 +70,7 @@ def requireAnomalyModel(func):
   """
   Decorator for functions that require anomaly models.
   """
+  @wraps(func)
```
Comment (rhyolight) [inline src/nupic/frameworks/opf/htm_prediction_model.py]: Without this, we lose the original docstrings on introspection, and Sphinx cannot get to them.

File: docs/source/api/opf/models.rst [hunk 2]
```diff
@@ -1,30 +1,50 @@
...
-   :members:
+   :members: getParameter,getRuntimeStats
    :show-inheritance:
-=======
+
+   .. automethod:: nupic.frameworks.opf.htm_prediction_model.HTMPredictionModel.setAnomalyParameter(param, value)
+   .. automethod:: nupic.frameworks.opf.htm_prediction_model.HTMPredictionModel.getAnomalyParameter(param)
+   .. automethod:: nupic.frameworks.opf.htm_prediction_model.HTMPredictionModel.anomalyRemoveLabels(start, end, labelFilter)
+   .. automethod:: nupic.frameworks.opf.htm_prediction_model.HTMPredictionModel.anomalyAddLabel(start, end, labelName)
+   .. automethod:: nupic.frameworks.opf.htm_prediction_model.HTMPredictionModel.anomalyGetLabels(start, end)
```
Comment (rhyolight) [inline docs/source/api/opf/models.rst]: I had to include these manually because they are decorated functions.

Comment (marionleborgne) [inline docs/source/api/opf/models.rst]: Weird. That's good to know

File: src/nupic/data/record_stream.py [hunk 3]
```diff
@@ -319,30 +329,42 @@ def getBookmark(self):
...
 
   def getFieldMin(self, fieldName):
-    """ Returns current minimum value for the field 'fieldName'.
-
+    """
     If underlying implementation does not support min/max stats collection,
     or if a field type does not support min/max (non scalars), the return
     value will be None.
+
+    :param fieldName: (string) name of field to get max
```
Comment (marionleborgne) [inline src/nupic/data/record_stream.py]: Should be: `to get min`

File: src/nupic/data/record_stream.py [hunk 4]
```diff
@@ -378,68 +402,93 @@ def clearStats(self):
...
 
   @abstractmethod
   def getFields(self):
-    """Returns a sequence of nupic.data.fieldmeta.FieldMetaInfo
-    name/type/special tuples for each field in the stream. Might be None, if
-    that information is provided externally (thru stream def, for example).
+    """
+    :returns: (list) of :class:`nupic.data.fieldmeta.FieldMetaInfo`objects for
+        each field in the stream. Might be None, if that information is provided
+        externally (thru stream def, for example).
```
Comment (marionleborgne) [inline src/nupic/data/record_stream.py]: Space missing between :class:`nupic.data.fieldmeta.FieldMetaInfo`  and `objects`, so the formatting is off in the HTML.

Super minor: consider spelling `thru` -> `through`.

File: src/nupic/data/record_stream.py [hunk 5]
```diff
@@ -378,68 +402,93 @@ def clearStats(self):
...
+    """
+    :returns: (int) index of ``category`` field
+    """
     return _getFieldIndexBySpecial(self.getFields(), FieldMetaSpecial.category)
 
 
   def getLearningFieldIdx(self):
-    """ Return index of the 'learning' field. """
+    """
+    :returns: (int) index of the ``learning ``field.
```
Comment (marionleborgne) [inline src/nupic/data/record_stream.py]: space is in the wrong spot. Should be  ``learning`` field.

File: src/nupic/frameworks/opf/model_factory.py [hunk 6]
```diff
@@ -56,11 +56,16 @@ def __getLogger(cls):
...
     """ Create a new model instance, given a description dictionary.
-    @param modelConfig (dict)
-           A dictionary describing the current model (TODO: schema)
-    @param logLevel (int) The level of logging output that should be generated
-    @exception (Exception) Unsupported model type
-    @returns (nupic.frameworks.opf.model.Model) A model.
+
+    :param modelConfig: (dict)
+           A dictionary describing the current model,
+           `described here <../../quick-start/example-model-params.html>`_.
```
Comment (marionleborgne) [inline src/nupic/frameworks/opf/model_factory.py]: Nice 👍

File: src/nupic/frameworks/opf/two_gram_model.py [hunk 7]
```diff
@@ -32,16 +32,17 @@
...
 
 class TwoGramModel(model.Model):
-  """Two-gram benchmark model."""
+  """
+  Two-gram benchmark model.
+
+  :param inferenceType: (:class:`nupic.frameworks.opf.opf_utils.InferenceType`)
+  :param encoders: a dict of dicts, eventually sent to
+         :meth:`~nupic.encoders.multi.MultiEncoder.addMultipleEncoders` (see
+         there for details).
```
Comment (marionleborgne) [inline src/nupic/frameworks/opf/two_gram_model.py]: Do you mean, see  method docs? (`addMultipleEncoders `) Or is there a missing hyperlink?

File: src/nupic/frameworks/opf/htm_prediction_model.py [hunk 8]
```diff
@@ -329,6 +342,9 @@ def getAnomalyParameter(self, param):
   def anomalyRemoveLabels(self, start, end, labelFilter):
     """
     Remove labels from the anomaly classifier within this model.
+
+    :param start: (int) index to start removing labels
+    :param end: (int) index to end removing labels
```
Comment (rhyolight) [inline src/nupic/frameworks/opf/htm_prediction_model.py]: shit I forgot labelFilter

File: src/nupic/frameworks/opf/two_gram_model.py (hunk 7 above)
Comment (rhyolight) [inline src/nupic/frameworks/opf/two_gram_model.py]: I meant see the `addMultipleEncoders` method docs for complete description of this parameter. Will make it clearer.

File: src/nupic/data/record_stream.py (hunk 3 above)
Comment (rhyolight) [inline src/nupic/data/record_stream.py]: I think this one is right.

**Comment by rhyolight** [comment]
I made changes in https://github.com/numenta/nupic/pull/3589/commits/1bf8894fe281d5ede8fb5c00bbcdab203f8c389c that don't seem to be reflected on the files page of this PR. Not sure why.

**Pull request merged**